stdlib/mutex.ml

```diff
@@ -18,3 +18,13 @@ external create: unit -> t = "caml_ml_mutex_new"
 external lock: t -> unit = "caml_ml_mutex_lock"
 external try_lock: t -> bool = "caml_ml_mutex_try_lock"
 external unlock: t -> unit = "caml_ml_mutex_unlock"
+
+(* cannot inline, otherwise flambda might move code around. *)
+let[@inline never] with_locked m f =
```

Comment:
You do not say why you need to avoid flambda reorderings. This is needed for async-exn-safety, but you do not mention other things needed to implement async-exn-safety. (If you want, feel free to adapt my version at https://gitlab.com/gadmm/memprof-limits/-/blob/master/src/mutex_aux.ml.)

Reply:
This is the exact same code, no? Why is yours async-exn-safe ?? How do you reason about such things if, as I understand it, _any_ call might poll and raise an exception??

Reply:
Yes, your implementation is async-exn-safe too. I was talking about documenting the code.

It is not true that any call might poll, the external call to `unlock` is an exception to this, this is why I had to declare it external instead of calling `Mutex.unlock`.

Reply:
And reasoning about polling locations is delicate indeed, it currently requires internal knowledge. This is why it is preferable to have it in the Stdlib, rather than letting users such as the Coq proof assistant rely on such internal knowledge (unfortunately the new polling attribute does not help in this case).

Reply:
That makes sense, thank you. We're saved by the fact that `Mutex.unlock` is a non allocating, non polling call to an external.